pkg/auth/providers/activedirectory/activedirectory_provider.go

```diff
@@ -164,16 +166,18 @@ func (p *adProvider) getActiveDirectoryConfig() (*v32.ActiveDirectoryConfig, *x5
 	storedADConfigMap := u.UnstructuredContent()
 
 	storedADConfig := &v32.ActiveDirectoryConfig{}
-	mapstructure.Decode(storedADConfigMap, storedADConfig)
-
-	metadataMap, ok := storedADConfigMap["metadata"].(map[string]interface{})
-	if !ok {
-		return nil, nil, fmt.Errorf("failed to retrieve ActiveDirectoryConfig metadata, cannot read k8s Unstructured data")
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Metadata:   nil,
+		DecodeHook: util.StringToK8sTimeHookFunc(),
+		Result:     storedADConfig,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create decoder for Active Directory Config: %w", err)
+	}
+	err = decoder.Decode(storedADConfigMap)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to decode Active Directory Config: %w", err)
```

Comment:
Instead of duplicating this code for every provider you an make a function in common. That does all this called decode.
``` golang
func Decode(input, output any) error {
    decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
		DecodeHook: stringToK8sTimeHookFunc(),
		Result:     output,
	})
	if err != nil {
		return nil, nil, fmt.Errorf("unable to create decoder for Config: %w", err)
	}
	err = decoder.Decode(input)
	if err != nil {
		return nil, nil, fmt.Errorf("unable to decode Config: %w", err)
	}
	return nil
}

Reply:
Yes, that makes perfect sense, good idea.